feat: add `math/base/special/hyp2f1`

anandkaranubc opened the pull request:
Resolves https://github.com/stdlib-js/stdlib/issues/122

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the JS implementation of the Gaussian hypergeometric function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/stdlib/issues/122

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: temp commit

anandkaranubc committed:
feat: initial files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: missing_dependencies
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
@kgryte, this is my first step into the big leagues... The PR is intentionally left incomplete, and only has the implementation and some basic tests. Since the implementation seems nearly finished (still need to test it further), I wanted to get your feedback and share my progress.

FYI, I used Joey's fixtures from his PR to test my implementation. Only four tests failed, but the test fixtures were relatively trivial. My next step is to create more refined tests to check every aspect of the code. Still a long way to go—haha!

anandkaranubc committed:
feat: update the JS implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
FYI: The `hyp2f1ra` function was moved to `hys2f1.js` because:

- It is only used in `hys2f1.js`.  
- Since both functions depend on each other, keeping them in separate JS files caused a TypeError, as `hys2f1` was not recognized as a function due to a circular dependency (MAIN REASON)

kgryte commented:
> The hyp2f1ra function was moved to hys2f1.js

Yes, this makes sense and is perfectly fine!

anandkaranubc committed:
test: update test fixtures

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: missing_dependencies
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
````diff
@@ -0,0 +1,256 @@
+* limitations under the License.
+*
+*
+* ## Notice
+*
+* The original C++ code and copyright notice are from the [SciPy library]{@link https://github.com/scipy/xsf/blob/main/include/xsf/cephes/hyp2f1.h}. The implementation has been modified for JavaScript.
+*
+* ```text
+* (C) Copyright SciPy Developers 2024.
+*
+* Use, modification and distribution are subject to the
+* SciPy Software License. (See accompanying file
+* LICENSE or copy at https://github.com/scipy/scipy/blob/main/LICENSE.txt)
+* ```
+*/
````
I’d appreciate some feedback on this and whether it should be included elsewhere as well.

kgryte replied:
Do we know how much that SciPy implementation differs from the original Cephes implementation?

kgryte replied:
But also, the license header should go in all relevant source files containing implementation code drawn from a reference implementation.

anandkaranubc replied:
Not at the moment. I can take a look at the original Cephes implementation.

kgryte replied:
If it is, say, 99% the same, I would argue that the license and attribution should be given to Cephes, not SciPy.

anandkaranubc replied:
I just checked the original [Cephes implementation](https://github.com/poliastro/cephes/blob/master/src/hyp2f1.c), and it is almost the same.

anandkaranubc replied:
I just checked the original [Cephes implementation](https://www.netlib.org/cephes/), and it is almost the same.

Right link this time :)

kgryte replied:
I'd use the Cephes license header then, as we do in other implementations which draw from Cephes.

anandkaranubc committed:
chore: fix linting issues

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: missing_dependencies
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feat/hyp2f1

anandkaranubc commented:
Is there any way to lint python files automatically?

anandkaranubc committed:
chore: fix python linting issues

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: missing_dependencies
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/utils.js:
```diff
@@ -0,0 +1,72 @@
+// EXPORTS //
+
+module.exports = {
+	'isNonPositiveInteger': isNonPositiveInteger,
```
@anandkaranubc In general, we prefer having one function/class per file. In which case, these should be modularized accordingly. If you were concerned about duplicating `EPS`, well, that can also go in a separate file and then be imported by the two modules. Yes, this is more files, but follows project conventions.

kgryte replied:
It is also fine to simply duplicate `EPS` in both modules.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,215 @@
+
+	if ( ( c < 0.0 && a <= c ) || ( c >= 0.0 && a >= c ) ) {
+		da = round( a - c );
+	}
+	else {
```
```suggestion
	} else {
```

In general, prefer a single line, unless you need to add a comment before the `else` clause.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,215 @@
+			t -= 1.0;
+		}
+	}
+	else {
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,215 @@
+		f1 = hys2f1( t, b, c, x, err );
+		loss += f1.error;
+		err = f1.error;
+		f0 = hys2f1( t + 1.0, b, c, x, err ); // CHECK IF err is THE SAME OR NEW ONE
```
```suggestion
		f0 = hys2f1( t + 1.0, b, c, x, err ); // CHECK IF err is THE SAME OR NEW ONE
```

Is this a TODO or an informative comment?

anandkaranubc replied:
Sorry, added it when I was testing it. I will remove that.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,215 @@
+		t += 1.0;
+		f1Val = f1.value;
+		f0Val = f0.value;
+		for ( n = 1; n < da; ++n ) {
```
```suggestion
		for ( n = 1; n < da; n++ ) {
```

We prefer post-fix unless there is reason not to.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hyt2f1.js:
```diff
@@ -0,0 +1,225 @@
+			y = hys2f1( a, c - b, c, -x / s, err );
+			val = pow( s, -a ) * y.value;
+		}
+		else {
```
Same comment as elsewhere.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+* // returns ~27.699
+*
+* v = hyp2f1( 3.0, 4.0, 7.0, 1.0 );
+* // returns PINF
```
```suggestion
* // returns +Infinity
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,256 @@
+
+// VARIABLES //
+
+var ETHRESH = 1.0e-12;
```
For these constants which get repeated, I'd move all of them to a single JSON file (e.g., `./config.json`) and then import them as needed. E.g.,

```
var ETHRESH = config.ETHRESH;
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,256 @@
+		y = hyt2f1( a, b, c, x, err );
+		return y.value;
+	}
+
```
Based on your readings of the literature, do you have any insights into the meaning/intent of the various branches here and throughout this implementation?

This isn't a requirement. I only ask because we typically try to include source comments explaining source implementations, where possible. This is a complex enough function that it may not be possible.

anandkaranubc replied:
There are indeed some areas where I can add short "explanatory" comments, following the source implementation. Just to confirm, is there a preference for placing those comments at the top of the code or within the relevant sections where they seem most appropriate?

kgryte replied:
Either is fine. We've typically done at the top of the code if we can explain comprehensively (e.g., `erf` functions). Otherwise, inline with the source.

anandkaranubc replied:
Mostly, these are "optimizations" by dividing the implementation into different cases, and trying different approximations. I will try my best to explain them comprehensively.

kgryte replied:
Feel free to just inline to start, and we can always pull out into the JSDoc if there is a consistent story which emerges.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js:
```diff
@@ -0,0 +1,295 @@
+	t.end();
+});
+
+tape( 'returns `1` if either `a` or `b` is 0 and `c` is not 0', function test( t ) {
```
```suggestion
tape( 'the function returns `1` if either `a` or `b` is 0 and `c` is not 0', function test( t ) {
```

Applies here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js:
```diff
@@ -0,0 +1,295 @@
+
+	for ( i = 0; i < x.length; i++ ) {
+		v = hyp2f1( a[ i ], b[ i ], c[ i ], x[ i ] );
+		if (expected[ i ] === 'PINF') {
```
Spacing. Applies here and elsewhere throughout this file.

kgryte commented:
Overall, this is looking good. No major red flags on my end, so should be ready for docs and benchmarks.

anandkaranubc committed:
refactor: apply suggestions from PR review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add docs, benchmarks and examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add inline comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc commented:
I've added the inline comments, most of which refer to/derive from the transformation formulas from [AMS55](https://www.convertit.com/go/convertit/reference/ams55.asp?Res=150&Page=559) and the original [Cephes](https://www.netlib.org/cephes/) implementation.

anandkaranubc committed:
feat: add inline comments #2

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/README.md:
```diff
@@ -0,0 +1,162 @@
+
+* * *
+
+## See Also
```
For a new package, these will get blown away, as the related package section is automatically generated by a centralized project database. As such, make a mental note, and after this PR is merged, we can add the function to the global namespace.

kgryte replied:
Sorry. I was not explicit. This list and the associated links should be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/README.md:
```diff
@@ -0,0 +1,162 @@
+
+<!-- <related-links> -->
+
+[@stdlib/math/base/special/gamma]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/gamma
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/README.md:
```diff
@@ -0,0 +1,162 @@
+
+<section class="intro">
+
+The [Hypergeometric function][hypergeometric-function] is defined for `|x| < 1` by the power series:
```
This intro is good and what we'd expect.

kgryte replied:
```suggestion
The [Gaussian hypergeometric function][hypergeometric-function] is defined for `|x| < 1` by the power series:
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/docs/types/test.ts:
```diff
@@ -0,0 +1,76 @@
+	hyp2f1( 2.0, 3.0, 5.0, 0.99 ); // $ExpectType number
+}
+
+// The compiler throws an error if the function is provided a first argument other than a number...
```
Rather than "other than a number" can we replace with "which is a not a number" here and below?

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,235 @@
+/**
+* Evaluates the Gaussian hypergeometric function by two-term recurrence in `a`.
+*
+* This helps prevent losing accuracy in the highly alternating hypergeometric series and allows a and b to be reduced to smaller values.
+*
+* AMS55 #15.2.10
```
```suggestion
* ## Notes
*
* -   This function helps prevent losing accuracy in the highly alternating hypergeometric series and allows a and b to be reduced to smaller values.
*
* ## References
*
* -   AMS55 #15.2.10
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hyt2f1.js:
```diff
@@ -0,0 +1,242 @@
+// MAIN //
+
+/**
+* Applies transformations for |x| near 1 then call the power series expansion.
```
```suggestion
* Applies transformations for `|x|` near unity before performing a power series expansion.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hyt2f1.js:
```diff
@@ -0,0 +1,242 @@
+			'error': err
+		};
+	}
+	// Use defining power series if no special cases
```
```suggestion
	// Perform a power series expansion if no special cases:
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return 1.0;
+	}
+
+	/* The transformation for c - a or c - b negative integer
+	* AMS55 #15.3.3
+	*/
```
```suggestion
	// The transformation for c - a or c - b negative integer (AMS55 #15.3.3):
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return pow( s, d ) * hyp2f1( c - a, c - b, c, x );
+	}
+
+	// The series diverges
```
```suggestion
	// Check whether the series diverges...
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	if ( negIntC ) {
+		// Check if termination before explosion
```
```suggestion
		// Check if termination before explosion...
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return PINF;
+	}
+
+	// Function is a polynomial, so try the power series expansion
```
```suggestion
	// Check whether the function is a polynomial before trying the power series expansion...
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return y.value;
+	}
+
+	/**
+	* This transform has a pole for integer b - a and may cause large cancellation errors near |1/x| = 1.
+	* AMS55 #15.3.7
+	*/
```
```suggestion
	// The following transform has a pole for integer b - a and may cause large cancellation errors near |1/x| = 1 (AMS55 #15.3.7)...
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return ( s * p ) + ( y * q );
+	}
+	if ( x < -1.0 ) {
+		// Transformation based on AMS55 #15.3.4
```
```suggestion
		// Transformation based on AMS55 #15.3.4:
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		if ( abs( a ) < abs( b ) ) {
+			return pow( s, -a ) * hyp2f1( a, c - b, c, x / ( x - 1.0 ) );
+		}
+		// Transformation based on AMS55 #15.3.5
```
```suggestion
		// Transformation based on AMS55 #15.3.5:
```

Okay. General rules for line comments:

-   If the comment applies to the next 1-3 lines without conditionals, end with a `:`.
-   If the comment applies to a block of code (e.g., within an `if` block or a `for` loop), use `...`.

Why the distinction?

-   The former is intended to convey that a comment applies only to what immediately follows.
-   The latter is intended to convey that a comment applies more generally to a chunk of code as a whole.

This is a minor quibble, but to be consistent, it would be good to "clean" up the comments in the source implementation accordingly.

anandkaranubc replied:
On it!

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+			r = e - 1.0;
+
+			// eslint-disable-next-line max-len
+			y = (((e * (r - (((2.0 * e) - q) * x))) * d2) + (((e - a) * (e - b) * x) * d1)) / (e * r * s);
```
```suggestion
			y = (((e * (r - (((2.0 * e) - q) * x))) * d2) + (((e - a) * (e - b) * x) * d1)) / (e * r * s);
```

General comment: how do we (I) think about whitespace?

- Use negative space to draw attention to what matters and to group related symbols.

E.g.,

```
x[ i%len ]
```

Why no spaces around `%`? Because, in general, it is useful to indicate that `i%len` is a single "unit". Namely, the operation which results in an index.

Why spaces in `x[ ]`? Because `x[i%len]` is harder to immediately grok where a variable ends and another begins. In `x[i%len]`, there are two pieces of salient info: 1) the array as conveyed by `x[` and 2) the indexing operation. Negative space can help delineate those two things.

So, for the above equation, what are the salient bits of info? Right now, you use whitespace to group parentheses and highlight operators. Maybe that is the right choice, but you could also use whitespace to group related operations/terms. E.g., `( e*r*s )`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js:
```diff
@@ -0,0 +1,285 @@
+		return y;
+	}
+
+	/* The transformation for c - a or c - b negative integer
```
Prefer a line comment.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js:
```diff
@@ -0,0 +1,295 @@
+	t.end();
+});
+
+tape( 'the function returns `PINF` when `c <= a + b`, `x === 1`, and neither `a` nor `b` are non-positive integers', function test( t ) {
```
```suggestion
tape( 'the function returns `PINF` when `c <= a + b`, `x === 1`, and neither `a` nor `b` are nonpositive integers', function test( t ) {
```

kgryte replied:
We don't do `non-negative`, so we similarly don't do `non-positive` despite what a spellchecker says.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/test/test.js:
```diff
@@ -0,0 +1,295 @@
+		delta = abs( v - expected[ i ] );
+
+		/*
+		* NOTE: The tolerance is set high in this case due to:
+		* 1. The expected values being either very small or very large.
+		* 2. The function making a large number of internal calls, leading to accumulated numerical errors.
```
```suggestion
		* NOTE: the tolerance is set high in this case due to:
		*
		* 1. The expected values having a very large range, being either very small or very large.
		* 2. The function making a large number of internal calls, leading to accumulated numerical errors.
```

anandkaranubc committed:
refactor: apply suggestions from PR review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hys2f1.js:
```diff
@@ -0,0 +1,239 @@
+			};
+		}
+		m = k + 1.0;
+		u *= ( ( f+k ) * ( g+k ) * x / ( ( h+k ) * m ) );
```
Yes, this is nice and makes clear the grouping.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/hyt2f1.js:
```diff
@@ -0,0 +1,242 @@
+			ax = ln( s );
+
+			// eslint-disable-next-line max-len
+			y = digamma( 1.0 ) + digamma( 1.0+e ) - digamma( a+d1 ) - digamma( b+d1 ) - ax;
```
Also nice use of grouping.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/README.md:
```diff
@@ -0,0 +1,162 @@
+
+# Hypergeometric function
+
+> Evaluates the [Hypergeometric function][hypergeometric-function].
```
```suggestion
> Evaluates the [Gaussian hypergeometric function][hypergeometric-function].
```

kgryte replied:
An uneducated question: is it "the Gaussian hypergeometric function" or is it "a Gaussian hypergeometric function"? Meaning, can be definitive that this is "the" one and only, or is this one of a family of associated functions?

If the latter, presumably, we'll have other variants.

kgryte replied:
Update: looked at the Wikipedia entry. Looks like this description is fine.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+/**
+* Evaluate the Gauss hypergoemetric function.
```
```suggestion
* Evaluate the Gaussian hypergoemetric function.
```

We should be consistent in our naming/descriptions.

kgryte replied:
Same question as posed in the README.

kgryte commented on lib/node_modules/@stdlib/math/base/special/hyp2f1/README.md:
```diff
@@ -0,0 +1,162 @@
+
+#### hyp2f1( a, b, c, x )
+
+Evaluates the [Hypergeometric function][hypergeometric-function].
```
```suggestion
Evaluates the [Gaussian hypergeometric function][hypergeometric-function].
```

anandkaranubc committed:
docs: remove links from README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: follow consistent naming

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte reviewed (approved):
LGTM. Thank you for working on this, @anandkaranubc!

anandkaranubc commented:
> LGTM. Thank you for working on this, @anandkaranubc!

Finally! The line I was longing to see.
